### PR TITLE
[multistage] adding support for range predicate

### PR DESCRIPTION
**Pull request by walterddr**
- adding rules to decompose range predicate into OR joined simple comparison. 
- if the look up is point or complemented points, we should still use IN / NOT IN
- if the search result in a constant value (true of false) it might trigger logical value literal node, which is also supported here.

### CODE REVIEW

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java [hunk 1]
```diff
@@ -200,6 +200,13 @@ private Object[][] provideTestSql() {
         new Object[]{"SELECT a.col1, b.col2 FROM a JOIN b ON a.col1 = b.col1 "
             + " WHERE a.col1 IN ('foo') AND b.col2 NOT IN ('')"},
 
+        // Range conditions with continuous and non-continuous range.
+        new Object[]{"SELECT a.col1, b.col2 FROM a JOIN b ON a.col1 = b.col1 "
+            + " WHERE a.col3 IN (1, 2, 3) OR (a.col3 > 10 AND a.col3 < 50)"},
```
Comment (ankitsultana) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java:205]: Can you also check the case where someone tries an impossible range condition or conditions with overlaps? I was running into an issue with those edge-cases in my PR. Examples:

```
(a.col3 > 10 AND a.col3 < 20) AND (a.col3 > 30 AND a.col3 < 40) # Impossible range
(a.col3 > 10 AND a.col3 < 20) AND (a.col3 > 15 AND a.col3 < 25) # Overlapping range
```

Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java:205]: yeah I was planning to split that into a separate PR but yeah sure let me create this one.

Comment (siddharthteotia) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java:205]: is IN (<single value>) getting rewritten to = ?

FWIW, in the current engine we do this rewrite but not really worth it

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/ValueNode.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+public class ValueNode extends AbstractStageNode {
+  @ProtoProperties
+  private List<List<RexExpression>> _literalRows;
+
+  public ValueNode(int stageId) {
+    super(stageId);
+  }
+
+  public ValueNode(int currentStageId, DataSchema dataSchema,
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/ValueNode.java:38]: General suggestion -- for every new operator / StageNode introduced, I think we should add a corresponding query compilation test to validate the plan, expressions etc. In the current engine, we have done that for quite some time in CalciteSqlCompilerTest.java

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java [hunk 3]
```diff
@@ -200,6 +200,16 @@ private Object[][] provideTestSql() {
...
+
+        new Object[]{"SELECT col1, SUM(col3) FROM a WHERE a.col3 BETWEEN 23 AND 36 "
+            + " GROUP BY col1 HAVING SUM(col3) > 10.0 AND MIN(col3) <> 123 AND MAX(col3) BETWEEN 10 AND 20"},
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java:208]: In the current engine non-literals (function expressions) on RHS get rewritten by moving to LHS. Do we support that here ?

`WHERE a > b + 20` -> `WHERE a - b > 20`

**Review by siddharthteotia — APPROVED** [review]

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java (hunk 1 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java:205]: single IN yes will be rewrite

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/ValueNode.java (hunk 2 above)
Comment (walterddr) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/ValueNode.java:38]: this one yes. `ValueNode` is a special node generated by calcite when the query evaluates back to a constant (regardless of underlying table) 

value node replaces table scan as a root.

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java (hunk 3 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java:208]: yes. this is not done by the range predicate but by the compilation on server.

**Pull request merged**